engine/access/rpc/backend/backend_stream_transactions.go

```diff
@@ -10,302 +10,240 @@ import (
-		// statuses are the same, the current transaction status should be retrieved.
-		txStatus, err = b.txLocalDataProvider.DeriveTransactionStatus(txInfo.blockWithTx.Height, txInfo.txExecuted)
+// checkBlockReady checks if the given block height is valid and available based on the expected block status.
+// Expected errors during normal operation:
+// - [subscription.ErrBlockNotReady]: block for the given block height is not available.
+func (b *backendSubscribeTransactions) checkBlockReady(height uint64) error {
+	// Get the highest available finalized block height
+	highestHeight, err := b.blockTracker.GetHighestHeight(flow.BlockStatusFinalized)
+	if err != nil {
+		return fmt.Errorf("could not get highest height for block %d: %w", height, err)
 	}
 
-	if err != nil {
-		if !errors.Is(err, state.ErrUnknownSnapshotReference) {
-			irrecoverable.Throw(ctx, err)
-		}
-		return flow.TransactionStatusUnknown, rpc.ConvertStorageError(err)
+	// Fail early if no block finalized notification has been received for the given height.
+	// Note: It's possible that the block is locally finalized before the notification is
+	// received. This ensures a consistent view is available to all streams.
+	if height > highestHeight {
+		return fmt.Errorf("block %d is not available yet: %w", height, subscription.ErrBlockNotReady)
 	}
 
-	return txStatus, nil
+	return nil
 }
 
-// generateResultsWithMissingStatuses checks if the current result differs from the previous result by more than one step.
+// generateResultsStatuses checks if the current result differs from the previous result by more than one step.
 // If yes, it generates results for the missing transaction statuses. This is done because the subscription should send
 // responses for each of the statuses in the transaction lifecycle, and the message should be sent in the order of transaction statuses.
 // Possible orders of transaction statuses:
 // 1. pending(1) -> finalized(2) -> executed(3) -> sealed(4)
 // 2. pending(1) -> expired(5)
 // No errors expected during normal operations.
-func (b *backendSubscribeTransactions) generateResultsWithMissingStatuses(
-	txInfo *transactionSubscriptionMetadata,
+func generateResultsStatuses(
+	txResult *access.TransactionResult,
 	prevTxStatus flow.TransactionStatus,
 ) ([]*access.TransactionResult, error) {
-	// If the previous status is pending and the new status is expired, which is the last status, return its result.
+	// If the old and new transaction statuses are still the same, the status change should not be reported, so
+	// return here with no response.
+	if prevTxStatus == txResult.Status {
+		return nil, nil
+	}
+
+	// If the previous status is pending or unknown and the new status is expired, which is the last status, return its result.
 	// If the previous status is anything other than pending, return an error, as this transition is unexpected.
```

Comment:
```suggestion
	// return immediately if the new status is expired, since it's the last status
	// If the previous status is anything other than pending or unknown, return an error since this transition is unexpected.
```